pkg/controller/common/container/defaulter.go

```diff
@@ -0,0 +1,175 @@
+	return *d.base.DeepCopy()
+}
+
+func NewDefaulter(base *corev1.Container) Defaulter {
+	return &defaulter{
+		base: base,
+	}
+}
+
+func (d defaulter) From(other corev1.Container) Defaulter {
+
+	if other.Lifecycle != nil {
+		d.WithPreStopHook(other.Lifecycle.PreStop)
+	}
+
+	return d.
+		WithImage(other.Image).
+		WithCommand(other.Command).
+		WithArgs(other.Args).
+		WithPorts(other.Ports).
+		WithEnv(other.Env).
+		WithResources(other.Resources).
+		WithVolumeMounts(other.VolumeMounts).
+		WithReadinessProbe(other.ReadinessProbe)
+}
+
+func (d defaulter) WithCommand(command []string) Defaulter {
+	if len(d.base.Command) == 0 {
+		d.base.Command = command
+	}
+	return d
+}
+
+func (d defaulter) WithArgs(args []string) Defaulter {
+	if len(d.base.Args) == 0 {
+		d.base.Args = args
+	}
+	return d
+}
+
+func (d defaulter) WithPorts(ports []corev1.ContainerPort) Defaulter {
+	for _, p := range ports {
+		if !d.portExists(p.Name) {
+			d.base.Ports = append(d.base.Ports, p)
+		}
+	}
+	return d
+}
+
+// portExists checks if a port with the given name already exists in the Container.
+func (d defaulter) portExists(name string) bool {
+	for _, p := range d.base.Ports {
+		if p.Name == name {
+			return true
+		}
+	}
+	return false
+}
+
+// WithImage sets up the Container Docker image, unless already provided.
+// The default image will be used unless customImage is not empty.
+func (d defaulter) WithImage(image string) Defaulter {
+	if d.base.Image == "" {
+		d.base.Image = image
+	}
+	return d
+}
+
+func (d defaulter) WithReadinessProbe(readinessProbe *corev1.Probe) Defaulter {
+	if d.base.ReadinessProbe == nil {
+		d.base.ReadinessProbe = readinessProbe
+	}
+	return d
+}
+
+// envExists checks if an env var with the given name already exists in the provided slice.
+func (d defaulter) envExists(name string) bool {
+	for _, v := range d.base.Env {
+		if v.Name == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (d defaulter) WithEnv(vars []corev1.EnvVar) Defaulter {
+	for _, v := range vars {
+		if !d.envExists(v.Name) {
+			d.base.Env = append(d.base.Env, v)
+		}
+	}
+	return d
+}
+
+func (d defaulter) WithResources(resources corev1.ResourceRequirements) Defaulter {
+	// Ensure resources are set
+	if d.base.Resources.Requests == nil && d.base.Resources.Limits == nil {
+		d.base.Resources = resources
+	}
+	return d
+}
+
+// volumeExists checks if a volume mount with the given name already exists in the Container.
+func (d defaulter) volumeMountExists(volumeMount corev1.VolumeMount) bool {
+	for _, v := range d.base.VolumeMounts {
+		if v.Name == volumeMount.Name || v.MountPath == volumeMount.MountPath {
+			return true
+		}
+	}
+	return false
+}
+
+func (d defaulter) WithVolumeMounts(volumeMounts []corev1.VolumeMount) Defaulter {
+	for _, v := range volumeMounts {
+		if !d.volumeMountExists(v) {
+			d.base.VolumeMounts = append(d.base.VolumeMounts, v)
+		}
+	}
+	// order volume mounts by name to ensure stable pod spec comparison
```

Comment:
not from your PR but since you're at it 😄 
we are sorting for stable comparison here, but we don't do it for environment variables or ports for example. I'm wondering if that sort is still necessary, or if we should also do it for other arrays.

Reply:
I think we should not sort env variable since they may reference each other, 👍 for the ports